lib/rust/profiler/macros/src/lib.rs

```diff
@@ -246,51 +242,74 @@ pub fn profile(
-    wrap_body(&orig_sig, *orig_body, &mut new_body);
-
-    func.block = new_body;
+    let obj_ident: syn::Ident = syn::parse(args)
+        .expect("The `profile` macro requires a profiling-level argument, e.g. #[profile(Task)]");
+    let label = make_label(&func.sig.ident);
+    match func.sig.asyncness {
+        Some(_) => profile_async(obj_ident, label, &mut func),
+        None => profile_sync(obj_ident, label, &mut func),
+    };
     func.into_token_stream().into()
 }
 
-// === instrument ===
-
-/// Insert instrumentation into the body, and genericize the profiler input to `impl Parent`.
-fn instrument(sig: &mut syn::Signature, body: &mut syn::Block) {
-    let profiler_label = make_label(&sig.ident);
-
-    // Parse profiler argument in signature.
-    let last_arg = match sig.inputs.last_mut() {
-        Some(syn::FnArg::Typed(pat_type)) => pat_type,
-        _ => panic!("Function annotated with #[profile] must have a profiler argument."),
+fn profile_async(obj_ident: syn::Ident, label: String, func: &mut syn::ItemFn) {
+    let start_profiler = start_profiler(obj_ident, label, func.sig.asyncness.is_some());
+    for s in &mut func.block.stmts {
+        WrapAwait.visit_stmt_mut(s);
+    }
+    let block = &func.block;
+    let body = quote::quote! {{
+        #start_profiler
+        async move {
+            profiler::Profiler::resume(&__profiler_scope.0);
+            let result = #block;
+            std::mem::drop(__profiler_scope);
+            result
+        }
+    }};
+    let output_ty = match &func.sig.output {
+        syn::ReturnType::Default => quote::quote! { () },
+        syn::ReturnType::Type(_, ty) => ty.to_token_stream(),
     };
-    let profiler = last_arg.pat.clone();
+    let output = quote::quote! {
+        -> impl std::future::Future<Output=#output_ty>
+    };
+    func.sig.asyncness = None;
+    func.sig.output = syn::parse2(output).unwrap();
+    func.block = syn::parse2(body).unwrap();
+}
 
-    // Replace profiler type `T `in sig with `impl Parent<T>`.
-    let profiler_type_in = last_arg.ty.clone();
-    let profiler_generic_parent = quote::quote! { impl profiler::Parent<#profiler_type_in> };
-    last_arg.ty = Box::new(syn::Type::Verbatim(profiler_generic_parent));
+fn profile_sync(obj_ident: syn::Ident, label: String, func: &mut syn::ItemFn) {
+    let start_profiler = start_profiler(obj_ident, label, func.sig.asyncness.is_some());
+    let block = &func.block;
+    let body = quote::quote! {{
+        #start_profiler
+        #block
+    }};
+    func.block = syn::parse2(body).unwrap();
+}
 
-    // Emit instrumentation statements.
-    let start_profiler = quote::quote! {
-        let _profiler = #profiler.new_child(#profiler_label);
-    };
-    let get_profiler = quote::quote! {
-        let #profiler = _profiler.profiler;
+fn start_profiler(
+    obj_ident: syn::Ident,
+    label: String,
+    asyncness: bool,
+) -> proc_macro2::TokenStream {
+    // TODO[kw] Optimization: pause_at(now)
```

Comment:
Please remove all todos. If you want to leave any TODO, you can do it, but:
1. If these are optimization TODOs, they can be described as implementation improvement, not as "TODO" (talk with other people on Discord what label we want to use there). "TODO" should mean a "missing functionality", not "optimization opportunity".
2. Such left "optimization opportunities" comments should be nicely described,so everyone reading the code should understand them.